src/server/routes/resetPassword.ts

```diff
@@ -90,6 +100,117 @@ router.get(
+
+		if (encryptedState?.email && encryptedState.stateHandle) {
+			const html = renderer('/reset-password/email-sent', {
+				requestState: mergeRequestState(state, {
+					pageData: {
+						email: readEncryptedStateCookie(req)?.email,
+						timeUntilTokenExpiry: convertExpiresAtToExpiryTimeInMs(
+							encryptedState.stateHandleExpiresAt,
+						),
+					},
+				}),
+				pageTitle: 'Check Your Inbox',
+			});
+			return res.type('html').send(html);
+		}
+		return res.redirect(
+			303,
+			addQueryParamsToPath('/reset-password', state.queryParams),
+		);
+	},
+);
+
+// handler for the passcode reset password form
+router.post(
+	'/reset-password/code',
+	handleAsyncErrors(async (req: Request, res: ResponseWithRequestState) => {
+		const { requestId } = res.locals;
+		const { code } = req.body;
+
+		const encryptedState = readEncryptedStateCookie(req);
+
+		// make sure we have the encrypted state cookie and the code otherwise redirect to the reset page
+		if (encryptedState?.stateHandle && code) {
+			const { stateHandle } = encryptedState;
+
+			try {
+				// submit the passcode to Okta
+				await submitPasscode({
+					passcode: code,
+					stateHandle,
+					introspectRemediation: 'challenge-authenticator',
+					challengeAnswerRemediation: 'reset-authenticator',
+					requestId,
+					ip: req.ip,
+				});
+
+				// update the encrypted state cookie to show the passcode was used
+				// so that if the user clicks back to the email sent page, they will be shown a message
+				updateEncryptedStateCookie(req, res, {
+					passcodeUsed: true,
+				});
+
+				// redirect to the password page to set a password depending on if the are setting
+				// or resetting a password
+				const authState = getAuthorizationStateCookie(req);
+				const passwordPage = (() => {
+					switch (authState?.confirmationPage) {
+						case '/set-password/complete':
+							return '/set-password/password';
+						default:
+							return '/reset-password/password';
+					}
+				})();
+
+				// redirect to the password page
+				return res.redirect(
+					303,
+					addQueryParamsToPath(passwordPage, res.locals.queryParams),
+				);
+			} catch (error) {
+				// handle passcode specific error
+				handlePasscodeError({
+					error,
+					req,
+					res,
+					emailSentPage: '/reset-password/email-sent',
+					expiredPage: '/reset-password/expired',
+				});
+
+				// if we redirected away during the handlePasscodeError function, we can't redirect again
+				if (res.headersSent) {
+					return;
+				}
+			}
+		}
+
+		// if we reach this point, redirect back to the reset password page, as something has gone wrong
+		return res.redirect(
+			303,
+			addQueryParamsToPath('/reset-password', res.locals.queryParams),
+		);
+	}),
+);
+
+// Route to resend the email for passcode reset password
+// Essentially the same as POST /reset-password, just start the process again
+router.post(
+	'/reset-password/code/resend',
+	redirectIfLoggedIn,
```

Comment:
Do we have a way of preventing spam here? this endpoint feels a bit abusable!

Reply:
Ah we have recaptcha protection on the email sent page, and the functionality for resend on the non-passcode flow uses recaptcha for this.

So I'll add that in!

Reply:
Oh and it shouldn't have `redirectIfLoggedIn` anyway, as we do allow logged in users to reset their password, so if this was kept in, then users who were logged in that clicked resend would be redirected to the reauthenticate page, which is not what we want!